### PR TITLE
ScalametaTokenizer: group sequences of whitespace

### PR DESCRIPTION
**Pull request by kitbellew**
Add and use uber whitespace-token containers; these will contain contiguous sequences of horizontal whitespace, or sequences of newlines, ignoring trailing or embedded horizontal space.

### CODE REVIEW

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala [hunk 1]
```diff
@@ -609,25 +609,26 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
       def findFirstEOL(pos: Int): Int =
         if (pos > indentPos) -1 else if (tokens(pos).is[AtEOL]) pos else findFirstEOL(pos + 1)
       @tailrec
-      def hasBlank(pos: Int, hadEOL: Boolean): Boolean =
+      def hasBlank(pos: Int, hadEOL: Boolean = false): Boolean =
```
Comment (tgodzik) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala:612]: Do we actually use `hadEOL` anymore? It's always false now

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala [hunk 2]
```diff
@@ -609,25 +609,26 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
...
         if (pos > indentPos) false
         else tokens(pos) match {
+          case _: MultiNL => true
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala:615]: Ach, right! I missed it, thanks for explaning

**Review by tgodzik — APPROVED** [review]

**Comment by bjaglin** [comment]
This looks like a small breaking change for consumers of the tokenizer, since consecutive whitespaces are now collapsed into composite tokens.

[Scalafix rules are impacted by that change](https://github.com/scalacenter/scalafix/pull/2024#issuecomment-2267622608) since tokens are exposed at the document level and via each element of the tree without any indirection, so instead of trying to abstract that change away, I am considering signalling it via a breaking change bump.

Which got me wondering:
- In the interest of minimizing impact on Scalafix rule authors, are there at this stage any other upcoming breaking changes on the horizon?
- If not, do you think there would be a way to preserve the old behavior for a bit longer on the 4.9.x line? Probably not, as the changelog mentions "Support CRLF, multiple-whitespace sequences" so I assume it's a feature, not a side effect 😃

**Comment by kitbellew** [comment]
> This looks like a small breaking change for consumers of the tokenizer, since consecutive whitespaces are now collapsed into composite tokens.

Oh well, I didn't realize scalafix was looking at tokens as well.

> * In the interest of minimizing impact on Scalafix rule authors, are there at this stage any other upcoming breaking changes on the horizon?

We usually try to avoid breaking changes... that is, if we know they are breaking. But no, no plans.

> * If not, do you think there would be a way to preserve the old behavior for a bit longer on the 4.9.x line? Probably not, as the changelog mentions "Support CRLF, multiple-whitespace sequences" so I assume it's a feature, not a side effect 😃

@tgodzik Sometime ago, we were debating whether to add some parameters to ScalametaParser, perhaps this might be one case. I think we ended up thinking Dialect is enough, don't remember why.

**Comment by tgodzik** [comment]
I guess in this case a setting would make sense, though not sure how to sensibly add it. We can have a separate tokenized method maybe?

**Comment by bjaglin** [comment]
> We can have a separate tokenized method maybe?

Any chance you could give me some pointers so that I look at that to unblock the bump on Scalafix side? I went through the tokenisation code paths, but I am not sure where to start.

**Comment by tgodzik** [comment]
I guess we could just add another method to ScalametaTokenizer which will take options. We can then just use that explicitely from scalafix? If needed we can expose it later via the extension methods, but I think it should not be necessary right now.

**Comment by kitbellew** [comment]
@bjaglin let me do it in the next day or two, and release, if this is blocking you. I'm generally in the middle of fixing some bugs that i found by attempting to use scalafmt on the dotty codebase, but can complete the fixes in the next patch.

**Comment by kitbellew** [comment]
does scalafix use the tokenizer directly? or through the parser?

**Comment by bjaglin** [comment]
> does scalafix use the tokenizer directly? or through the parser?

Only through the parser.

**Pull request merged**